### PR TITLE
Add doc

### PR DESCRIPTION
**Pull request by dimriou**
[BCFR-881](https://smartcontract-it.atlassian.net/browse/BCFR-881)

[BCFR-881]: https://smartcontract-it.atlassian.net/browse/BCFR-881?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: core/chains/evm/gas/docs/FEE_HISTORY_ESTIMATOR.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
+# Fee History Estimator
+
+## Overview
+`Fee History` estimator is an EVM-based gas estimator that utilizes RPC calls to make gas price estimations. The estimator heavily relies on two RPC calls: `eth_gasPrice` and `eth_feeHistory`. It is built as a service and caches the calculated results in order to minimize overhead. It can bump past estimations by aggregating the result with the latest market prices. `Fee History` estimator supports both Legacy and Dynamic(EIP-1559) transactions. It can also handle chains that don't have a mempool and process transactions in an FCSF manner.
```
Comment (amit-momin) [inline core/chains/evm/gas/docs/FEE_HISTORY_ESTIMATOR.md]: I think this line can be a little confusing for someone without context: `It can bump past estimations by aggregating the result with the latest market prices.`

Comment (amit-momin) [inline core/chains/evm/gas/docs/FEE_HISTORY_ESTIMATOR.md]: Might be helpful to write this out for anyone that's not familiar.
```suggestion
`Fee History` estimator is an EVM-based gas estimator that utilizes RPC calls to make gas price estimations. The estimator heavily relies on two RPC calls: `eth_gasPrice` and `eth_feeHistory`. It is built as a service and caches the calculated results in order to minimize overhead. It can bump past estimations by aggregating the result with the latest market prices. `Fee History` estimator supports both Legacy and Dynamic(EIP-1559) transactions. It can also handle chains that don't have a mempool and process transactions on a first-come, first-serve (FCFS) basis.
```

File: core/chains/evm/gas/docs/FEE_HISTORY_ESTIMATOR.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+## Configs
+- `BumpPercent`: is the percentage by which to bump gas on a transaction. This is used when the estimator's bumping API gets called.
+- `CacheTimeout`: the time to wait to refresh the cached values stored. A small jitter is applied so the timeout won't be exactly the same each time.
```
Comment (amit-momin) [inline core/chains/evm/gas/docs/FEE_HISTORY_ESTIMATOR.md]: nit: just to be consistent with how the other definitions are written. Also `stored` might be redundant if we mention they're `cached`
```suggestion
- `CacheTimeout`: is the time to wait to refresh the cached values. A small jitter is applied so the timeout won't be exactly the same each time.
```

File: core/chains/evm/gas/docs/FEE_HISTORY_ESTIMATOR.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+- `BlockHistorySize`: controls the number of past blocks to include during gas calculations. If set to 0, the estimator will skip any priority fee calculation and calculate the underlying base fee. This config should be set to 0 for chains that don't have a mempool. 
+- `RewardPercentile`: it specifies which fee percentile to pick from for each processed past block. 
+
+### Validations
+During startup, the estimator will perform two config checks:
+- `BumpPercent` is equal to or higher than *MinimumBumpPercentage*. *MinimumBumpPercentage* is fixed at 10% and it's the minimum percentage allowed by Geth when bumping a transaction, to prevent spam attacks. Replacing a transaction with a price less than 10% from the previous one will result in an error on the RPC side. Even for chains that don't enforce that rule, a 10% bump seems reasonable.
+- `RewardPercentile` is equal or lower than *ConnectivityPercentile*, when `EIP1559` is enabled. *ConnectivityPercentile* is fixed at 85% and it's the maximum percentile we're willing to bump a transaction's price. This is used as a sanity check method in order to avoid excessive gas bumping when an RPC is not responding.
```
Comment (amit-momin) [inline core/chains/evm/gas/docs/FEE_HISTORY_ESTIMATOR.md]: nit: since this is technically a percentile
```suggestion
- `RewardPercentile` is equal or lower than *ConnectivityPercentile*, when `EIP1559` is enabled. *ConnectivityPercentile* is fixed at the 85th percentile and it's the maximum percentile we're willing to bump a transaction's price. This is used as a sanity check method in order to avoid excessive gas bumping when an RPC is not responding.
```

File: core/chains/evm/gas/docs/FEE_HISTORY_ESTIMATOR.md [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+## As a Service
+`Fee History` estimator is built as a service. This means it will periodically poll the RPC for new prices, perform off-chain calculations, and cache the result for future use. For simplicity, only one type of gas estimation can be enabled at a time, Legacy or Dynamic. The poll interval is controlled by `CacheTimeout`. This value should be close to the block time. For slower chains, like Ethereum, you can set it to 12s, the same as the block time. For faster chains, you can skip a block or two, as prices will be refreshed more frequently. Ideally, 1s should be the absolute minimum. 
+
+
+## Gas Price Estimations
```
Comment (amit-momin) [inline core/chains/evm/gas/docs/FEE_HISTORY_ESTIMATOR.md]: Just to differentiate this from dynamic more explicitly, could we rename this to `Legacy Gas Price Estimation`?
```suggestion
## Legacy Gas Price Estimations
```

File: core/chains/evm/gas/docs/FEE_HISTORY_ESTIMATOR.md [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+## Dynamic Price Estimations
+### Fetching
+`Fee History` estimator periodically calls `eth_feeHistory` method to get the most up-to-date information from the RPC (more information about the call can be found [here](https://ethereum.github.io/execution-apis/api-documentation/)). It fetches three things:
+- Base Fee of the next block
+- The Yth priority fee percentiles of the past X blocks, where Y is controlled by `RewardPercentile` and X by `BlockHistorySize`.
+- The 85th priority fee percentiles of the past X blocks.
+
+The above values are used to construct and cache the following:
+- **MaxPriorityFeePerGas**: the average of Yth priority fee percentiles, excluding zero values.
+- **MaxFeePerGas**: `baseFee * BaseFeeBufferPercentage + MaxPriorityFeePerGas`
```
Comment (amit-momin) [inline core/chains/evm/gas/docs/FEE_HISTORY_ESTIMATOR.md]: Think we should maybe briefly call out and explain what `BaseFeeBufferPercentage` is and why we have it. To someone that hasn't looked at the code, this could seem like it came out of nowhere

File: core/chains/evm/gas/docs/FEE_HISTORY_ESTIMATOR.md [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+- **PriorityFeeThreshold**: the max out of every 85th priority fee percentile. This value is used to stop the estimator from bumping a price above that threshold and represents the maximum allowed value.
+
+*Note*: for chains that don't have a mempool (activated with `BlockHistorySize=0`) **MaxPriorityFeePerGas** and **PriorityFeeThreshold** are set to 0 since there is no concept of gas bumping.
+
+### Bumping
+For bumping, `Fee History` estimator bumps both maxPriorityFeePerGas and maxFeePerGas of the original transaction attempt. This is required by Geth, along with the 10% minimum bumping threshold. The bumped attempt is aggregated with the cached market prices that are stored in the estimator and the one with the highest value is picked. Finally, the resulting maxPriorityFeePerGas gets compared with the cached PriorityFeeThreshold value. If the bumped value is higher than that, this indicates a potential connection issue with the RPC, and bumping is skipped, returning an error.
```
Comment (amit-momin) [inline core/chains/evm/gas/docs/FEE_HISTORY_ESTIMATOR.md]: I think `aggregated` might not be the right word here
```suggestion
For bumping, `Fee History` estimator bumps both maxPriorityFeePerGas and maxFeePerGas of the original transaction attempt. This is required by Geth, along with the 10% minimum bumping threshold. The bumped price is compared to the cached market prices stored in the estimator and the highest of the two is picked. Finally, the resulting maxPriorityFeePerGas gets compared to the cached PriorityFeeThreshold value. If the bumped value is higher, this indicates a potential connection issue with the RPC, and bumping is skipped, returning an error.
```

File: core/chains/evm/gas/docs/FEE_HISTORY_ESTIMATOR.md [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+*Note*: for chains that don't have a mempool (activated with `BlockHistorySize=0`) bumping works differently. Instead, we force-fetch the most up-to-date Base Fee value and embed it to MaxFeePerGas. MaxPriorityFeePerGas remains 0.
```
Comment (amit-momin) [inline core/chains/evm/gas/docs/FEE_HISTORY_ESTIMATOR.md]: nit:
```suggestion
*Note*: for chains that don't have a mempool (activated with `BlockHistorySize=0`) bumping works differently. Instead, we force-fetch the most up-to-date Base Fee value and embed it in the MaxFeePerGas. MaxPriorityFeePerGas remains 0.
```

**Review by amit-momin — CHANGES_REQUESTED** [review]

File: core/chains/evm/gas/docs/FEE_HISTORY_ESTIMATOR.md (hunk 1 above)
Comment (huangzhen1997) [reply core/chains/evm/gas/docs/FEE_HISTORY_ESTIMATOR.md]:
```suggestion
`Fee History` estimator is an EVM-based gas estimator that utilizes RPC calls to make gas price estimations. The estimator heavily relies on two RPC calls: `eth_gasPrice` and `eth_feeHistory`. It is built as a service and caches the calculated results in order to minimize overhead. It can bump past estimations by aggregating the result from most recent blocks. `Fee History` estimator supports both Legacy and Dynamic(EIP-1559) transactions. It can also handle chains that don't have a mempool and process transactions in an FCSF manner.
```

Comment (dimriou) [reply core/chains/evm/gas/docs/FEE_HISTORY_ESTIMATOR.md]: I think we already have a couple of these in the docs like EIP, EVM, RPC, etc. I don't expect non-technical people to read this.

Comment (dimriou) [reply core/chains/evm/gas/docs/FEE_HISTORY_ESTIMATOR.md]: I didn't want to get into too much detail here, since we explain FHE's behavior below and the code has plenty of comments. Is there something in particular that doesn't make sense?

Comment (amit-momin) [reply core/chains/evm/gas/docs/FEE_HISTORY_ESTIMATOR.md]: Maybe in that case we can omit this detail here and let people learn this below? This line is trying to say that if the market has spiked above the bumped value, the estimator would use the relevant market price instead right?

Comment (amit-momin) [reply core/chains/evm/gas/docs/FEE_HISTORY_ESTIMATOR.md]: Ok in that case maybe just the nit to fix the abbreviation to `FCFS` then

File: core/chains/evm/gas/docs/FEE_HISTORY_ESTIMATOR.md [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+## Configs
+- `BumpPercent`: is the percentage by which to bump gas on a transaction. This is used when the estimator's bumping API gets called.
+- `CacheTimeout`: is the time to wait to refresh the cached values. A small jitter is applied so the timeout won't be exactly the same each time.
+- `EIP1559`: enables EIP-1559 mode and deactivates Legacy estimations. This means the estimator will refresh prices and make estimations only for Dynamic transactions.
+
+The rest of the configs are only applicable when `EIP1559` is enabled
+
+- `BlockHistorySize`: controls the number of past blocks to include during gas calculations. If set to 0, the estimator will skip any priority fee calculation and calculate the underlying base fee. This config should be set to 0 for chains that don't have a mempool. 
+- `RewardPercentile`: it specifies which fee percentile to pick from for each processed past block.
```
Comment (amit-momin) [inline core/chains/evm/gas/docs/FEE_HISTORY_ESTIMATOR.md]: nit: just to keep these definition formats consistent
```suggestion
- `RewardPercentile`: specifies which fee percentile to pick from for each processed past block. 
```

File: core/chains/evm/gas/docs/FEE_HISTORY_ESTIMATOR.md [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+## Dynamic Price Estimations
+### Fetching
+`Fee History` estimator periodically calls `eth_feeHistory` method to get the most up-to-date information from the RPC (more information about the call can be found [here](https://ethereum.github.io/execution-apis/api-documentation/)). It fetches three things:
+- Base Fee of the next block
+- The Yth priority fee percentiles of the past X blocks, where Y is controlled by `RewardPercentile` and X by `BlockHistorySize`.
+- The 85th priority fee percentiles of the past X blocks.
+
+The above values are used to construct and cache the following:
+- **MaxPriorityFeePerGas**: the average of Yth priority fee percentiles, excluding zero values.
+- **MaxFeePerGas**: `baseFee * BaseFeeBufferPercentage + MaxPriorityFeePerGas`. BaseFeeBufferPercentage is used as a safety to catch any fluctuations in the Base Fee during the next blocks.
```
Comment (amit-momin) [inline core/chains/evm/gas/docs/FEE_HISTORY_ESTIMATOR.md]: nit: Since this is referring to a variable name, should we mark it with `
```suggestion
- **MaxFeePerGas**: `baseFee * BaseFeeBufferPercentage + MaxPriorityFeePerGas`. `BaseFeeBufferPercentage` is used as a safety to catch any fluctuations in the Base Fee during the next blocks.
```

File: core/chains/evm/gas/docs/FEE_HISTORY_ESTIMATOR.md (hunk 1 above)
Comment (dimriou) [reply core/chains/evm/gas/docs/FEE_HISTORY_ESTIMATOR.md]: I wanted to quickly mention that bumping also takes into account market prices but not go into too much detail because then I would have to mention the caps and the limits. If this seems confusing, I can remove it instead.

File: core/chains/evm/gas/docs/FEE_HISTORY_ESTIMATOR.md (hunk 9 above)
Comment (dimriou) [reply core/chains/evm/gas/docs/FEE_HISTORY_ESTIMATOR.md]: I used italics for const variables above, so I'll use the same here.

File: core/chains/evm/gas/docs/FEE_HISTORY_ESTIMATOR.md [hunk 10]
```diff
@@ -0,0 +1,56 @@
+# Fee History Estimator
+
+## Overview
+`Fee History` estimator is an EVM-based gas estimator that utilizes RPC calls to make gas price estimations. The estimator heavily relies on two RPC calls: `eth_gasPrice` and `eth_feeHistory`. It is built as a service and caches the calculated results in order to minimize overhead. It can bump past estimations by aggregating the result with the latest market prices. `Fee History` estimator supports both Legacy and Dynamic(EIP-1559) transactions. It can also handle chains that don't have a mempool and process transactions on an FCSF basis.
```
Comment (amit-momin) [inline core/chains/evm/gas/docs/FEE_HISTORY_ESTIMATOR.md]: nit:
```suggestion
`Fee History` estimator is an EVM-based gas estimator that utilizes RPC calls to make gas price estimations. The estimator heavily relies on two RPC calls: `eth_gasPrice` and `eth_feeHistory`. It is built as a service and caches the calculated results in order to minimize overhead. It can bump past estimations by aggregating the result with the latest market prices. `Fee History` estimator supports both Legacy and Dynamic(EIP-1559) transactions. It can also handle chains that don't have a mempool and process transactions on an FCFS basis.
```

File: core/chains/evm/gas/docs/FEE_HISTORY_ESTIMATOR.md (hunk 1 above)
Comment (amit-momin) [reply core/chains/evm/gas/docs/FEE_HISTORY_ESTIMATOR.md]: I see. In that case maybe rewording might be better? Something like: `While bumping, it prioritizes using the latest result from most recent blocks if it exceeds the bumped gas price`. What do you think?

**Review by amit-momin — APPROVED** [review]

File: core/chains/evm/gas/docs/FEE_HISTORY_ESTIMATOR.md [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+## Configs
+- `BumpPercent`: is the percentage by which to bump gas on a transaction. This is used when the estimator's bumping API gets called.
+- `CacheTimeout`: is the time to wait to refresh the cached values. A small jitter is applied so the timeout won't be exactly the same each time.
+- `EIP1559`: enables EIP-1559 mode and deactivates Legacy estimations. This means the estimator will refresh prices and make estimations only for Dynamic transactions.
+
+The rest of the configs are only applicable when `EIP1559` is enabled
+
+- `BlockHistorySize`: controls the number of past blocks to include during gas calculations. If set to 0, the estimator will skip any priority fee calculation and calculate the underlying base fee. This config should be set to 0 for chains that don't have a mempool. 
+- `RewardPercentile`: specifies which fee percentile to pick from for each processed past block.
```
Comment (simsonraj) [inline core/chains/evm/gas/docs/FEE_HISTORY_ESTIMATOR.md:14]: is there a default for this? if so can we have 1 liner about why we chose that as a default

**Review by simsonraj — APPROVED** [review]
